### PR TITLE
#1339 Add API for bus connected by switch

### LINKED ISSUE #1339: Service in data interface to check switches in bus breaker case

### PR DESCRIPTION
**Pull request by lecourtoisflo**
closes #1339

### CODE REVIEW

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp [hunk 1]
```diff
@@ -415,4 +415,10 @@ VoltageLevelInterfaceIIDM::mapConnections() {
   }
 }
 
+std::vector<std::string>
+VoltageLevelInterfaceIIDM::getBusesConnectedBySwitch(const std::string&) {
+  // this service is not supported by IIDM 1.0
```
Comment (marcochiaramello) [inline dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp:420]: I'd add a dynawo log notifying the user he can not use it for iidm 1.0

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h [hunk 2]
```diff
@@ -215,6 +215,16 @@ class VoltageLevelInterface {
...
+
+  /**
+   * @brief Retrieve the buses linked to a bus by a switches network path
+   *
+   * This Service API can be used only if topology is bus breaker
+   *
+   * @param busId bus id to check
+   * @returns the list of the bus ids linked to @p busId
+   */
+  virtual std::vector<std::string> getBusesConnectedBySwitch(const std::string& busId) = 0;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h:227]: Before reviewing the algorithm: I'm not sure putting those services inside the VL interface is the best way to do this. This is very specific to DynaFlow, and not used in the core of the simulation. Wouldn't it be possible to have a service manager on the side to perform this?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h (hunk 2 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h:227]: @adriguir any opinion?

Comment (marcochiaramello) [reply dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h:227]: I think this is a good idea

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp (hunk 1 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp:420]: DEBUG ? If it is warning or error, I'm not sure since it is a feature for further version

Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp:420]: should be a throw, and if we change the architecture on this it won't be needed anymore

Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp:420]: (there is not future for the IIDM folder)

File: dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h (hunk 2 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h:227]: We already have a similar service in the case of node breaker besides this new one. If we create a service manager, we should also move ALL services like this one in it. Seems a lot heavy to me

Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNVoltageLevelInterface.h:227]: which service are we referring to here? If it is the graph exploration for node braker case it is not really a service as it is required for the core to work

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp (hunk 1 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNVoltageLevelInterfaceIIDM.cpp:420]: This case cannot happen in IIDM1.0 so we consider that we are always in the case "2 buses cannot be linked by switches" in this case. Comment will be updated

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+void
+ServiceManagerInterfaceIIDM::buildGraph(const boost::shared_ptr<VoltageLevelInterface>& vl) {
+  auto compare = [](const boost::shared_ptr<BusInterface>& bus, const std::string& id) { return bus->getID() == id; };
+
+  auto& graph = graphs_[vl->getID()];
+  const auto& buses = vl->getBuses();
+
+  for (size_t i = 0; i < buses.size(); i++) {
+    // we are using the position in the vector as the vertex, as these buses are fixed
+    graph.addVertex(i);
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:40]: couldn't we create a dictionary here to avoid  using find_if below?

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  // Change weight of edges
+  boost::unordered_map<std::string, float> weights;
+  for (const auto& sw : (*vlIt)->getSwitches()) {
+    weights[sw->getID()] = sw->isOpen() ? 0 : 1;
+  }
+
+  std::vector<std::string> ret;
+
+  size_t busIndexFound = it - buses.begin();
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:93]: why using an iterator instead of directly comparing the id within the loop?

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  for (size_t busIndex = 0; busIndex < buses.size(); busIndex++) {
+    if (busIndex == busIndexFound) {
+      continue;
+    }
+    if (graphs_.at((*vlIt)->getID()).pathExist(busIndexFound, busIndex, weights)) {
+      ret.push_back(buses.at(busIndex)->getID());
+    }
+  }
+
+  // Bus ids are unique so we can stop the loop here
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:103]: Is this comment really useful?

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * @brief Implementation for IIDM dynawo service
+ */
+class ServiceManagerInterfaceIIDM : public ServiceManagerInterface {
+ public:
+  /**
+   * @brief Constructor
+   *
+   * @param dataInterface data interface corresponding to current service manager
+   */
+  explicit ServiceManagerInterfaceIIDM(const DataInterfaceIIDM* const dataInterface);
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.h:45]: Please used shared_ptr

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  vl.addBus(bus2);
+  vl.addBus(bus3);
+  vl.addBus(bus4);
+
+  interface.initFromIIDM();
+
+  auto serviceManager = interface.getServiceManager();
+
+  try {
+    serviceManager->getBusesConnectedBySwitch("BUS0", vl.getID());
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp:114]: Please use ASSERT_THROW_DYNAWO

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h [hunk 8]
```diff
@@ -155,6 +156,13 @@ class DataInterface {
    * @param filepath file to create
    */
   virtual void dumpToFile(const std::string& filepath) const = 0;
+
+  /**
+   * @brief Retrieve service manager associated with this data interface
+   *
+   * @returns the service manager
+   */
+  virtual boost::shared_ptr<ServiceManagerInterface> getServiceManager() const = 0;
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:165]: irrelevant in the case of a pointer

File: dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+#ifndef MODELER_DATAINTERFACE_DYNSERVICEMANAGERINTERFACE_H_
+#define MODELER_DATAINTERFACE_DYNSERVICEMANAGERINTERFACE_H_
+
+#include <string>
+#include <vector>
+
+namespace DYN {
+
+/**
+ * @brief Interface for dynwo service interface
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/DYNServiceManagerInterface.h:29]: OK

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.cpp [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * @brief Service manager : implementation file for IIDM implementation
+ *
+ */
+
+#include "DYNServiceManagerInterfaceIIDM.h"
+
+namespace DYN {
+std::vector<std::string>
+ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string&, const std::string&) {
+  // IIDM 1.0 doesn't support buses connected by switches so we alwayrs return an empty array
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNServiceManagerInterfaceIIDM.cpp:26]: OK

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 3 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:40]: Using this way allows to reuse DYN graph component, which is much easier and avoids code duplicata

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:93]: Dyn Graph is using indexes so we use indexes as well to avoid computing the distance at each step of the loop

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:165]: ok but then we should remove the const as the ServiceManagerInterface might be modified after this call

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:103]: No, artefact from previous implementation. will be removed

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 3 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:40]: yes, I mean a dictionary bus-id=> index to avoid searching several time the index of each bus into the vector

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.h (hunk 6 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.h:45]: Not in this case. Shared_ptr have ownership of the pointer, which we do not want here. Besides, ServiceManagerInterfaceIIDM constructor is called inside DataInterfaceIIDM interface.

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 4 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:93]: yes but we go though the buses just below so we have a direct access to their ids

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp (hunk 7 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestServiceManagerInterface.cpp:114]: OK

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:93]: ids, yes not their index

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 3 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:40]: OK

File: dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp (hunk 4 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp:93]: Let me rephrase: buses contain the BusInterface. Above you search which iterator corresponds to bus->getID() == busId, and then you compare this index to busIndex.
Why not replacing all of this by if (buses[busIndex]->getID() == busId) that would be way more readable?

**Review by rosiereflo — APPROVED** [review]

**Review by marcochiaramello — APPROVED** [review]

**Review by marcochiaramello — CHANGES_REQUESTED** [review]
@lecourtoisflo It seems there is a code smell, can you fix it?

**Comment by lecourtoisflo** [comment]
> @lecourtoisflo It seems there is a code smell, can you fix it?

It is a false positive

> @lecourtoisflo It seems there is a code smell, can you fix it?

It is a false positive made for people coming for C world.

**Comment by rosiereflo** [comment]
it does raise a question though: why using a struct and not a class?

**Comment by lecourtoisflo** [comment]
> it does raise a question though: why using a struct and not a class?

If there is no invariant (i.e. no private member) there is no rational for using class, because struct put all public by default

**Comment by rosiereflo** [comment]
yes I agree with this, I've marked the code smell as won't fix

**Review by marcochiaramello — APPROVED** [review]

**Review by gautierbureau — APPROVED** [review]
